### PR TITLE
feat: add data use restriction column to HCA Data Explorer (#4136)

### LINKED ISSUE #4136: [HCA DX] Add data use restriction facet and column
### Background
See [Azul ticket](https://github.com/DataBiosphere/azul/issues/6131).

### Need
Using the `projects.dataUseRestriction` value returned from Azul, corresponding columns and facets are to be displayed as follows.

#### Data use restriction column
- "Projects" entity after the "Access" column
- "Files" entity after the "File Size" column
 
#### Data use restriction filter
- "Project" section after the "Biological Network" filter

### Approach
* Convert `null` values to "".

### PR DESCRIPTION
**Pull request by mrtopsyt**
### Ticket

Closes #4136 .

### Reviewers

@MillenniumFalconMechanic .

### Changes

- Added "Data Use Restriction" column to the Projects and Files entity on the Managed Access config
- Added "Data Use Restriction" term facet to the Managed Access config

#### Notes
Note that the projects.dataUseRestriction endpoint contains only null values on the dev backend, so in order to see any values for this column it is necessary to use the prod site

### CODE REVIEW

File: app/common/types.ts
Comment (frano-m) [inline app/common/types.ts:1]: @jpaten I think we should revert this change and update the type in `site-config/hca-dcp/cc-ma-dev/index/common/constants.ts` to:

```
export const COLUMN: PickSome<
  Record<keyof typeof HCA_DCP_CATEGORY_KEY, ColumnConfig>,
  "ACCESSIBLE" | "DATA_USE_RESTRICTION"
> = {}
```

File: app/viewModelBuilders/azul/hca-dcp/common/viewModelBuilders.ts [hunk 1]
```diff
@@ -717,6 +717,23 @@ export const buildDataCurators = (
...
+ * @returns model to be used as props for the BasicCell component
+ */
+export const buildDataUseRestriction = (
+  projectsResponse: ProjectsResponse
+): React.ComponentProps<typeof C.BasicCell> => {
+  return {
+    value: processEntityValue(
+      projectsResponse.projects,
+      "dataUseRestriction",
+      LABEL.EMPTY
```
Comment (frano-m) [inline app/viewModelBuilders/azul/hca-dcp/common/viewModelBuilders.ts:750]: Let's remove `LABEL.EMPTY` - the fallback value for `processEntityValue` is `LABEL.UNSPECIFIED`; any `null` values returned will be displayed as "Unspecified" within the table.

Comment (frano-m) [reply app/viewModelBuilders/azul/hca-dcp/common/viewModelBuilders.ts:750]: image

File: site-config/hca-dcp/cc-ma-dev/index/common/constants.ts
Comment (frano-m) [inline site-config/hca-dcp/cc-ma-dev/index/common/constants.ts:1]: Change to:

```
export const COLUMN: PickSome<
  Record<keyof typeof HCA_DCP_CATEGORY_KEY, ColumnConfig>,
  "ACCESSIBLE" | "DATA_USE_RESTRICTION"
> = {}
```

File: site-config/hca-dcp/cc-ma-dev/index/filesEntityConfig.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  entitiesConfig: EntityConfig
+): EntityConfig {
+  // Clone files entity.
+  const cloneEntity = { ...entitiesConfig };
+  // Clone list.
+  const cloneList = { ...cloneEntity.list };
+  // Clone columns.
+  const cloneColumns = [...cloneList.columns];
+  // Add data use restriction column.
+  cloneColumns.splice(2, 0, COLUMN.DATA_USE_RESTRICTION);
```
Comment (frano-m) [inline site-config/hca-dcp/cc-ma-dev/index/filesEntityConfig.ts]: The shape returned for `files`'s `project` is different to `projects`'s `project` response. e.g.

```
dataUseRestriction: [null]
```

1. I believe you will need to add to [AggregatedProjectResponse](https://github.com/DataBiosphere/data-browser/blob/c4335416e8d0f456e58f15d8419df5ab616ec309/app/apis/azul/hca-dcp/common/aggregatedEntities.ts#L80) `dataUseRestriction: (null | string)[]`.
2. On line 19, remove the reference `COLUMN.DATA_USE_RESTRICTION` because this is `projects` response related. I think for now let's just insert the new config like this:

```
  cloneColumns.splice(2, 0, {
    componentConfig: {
      component: C.BasicCell,
      viewBuilder: V.buildAggregatedDataUseRestriction,
    } as ComponentConfig<typeof C.BasicCell>,
    header: HCA_DCP_CATEGORY_LABEL.DATA_USE_RESTRICTION,
    id: HCA_DCP_CATEGORY_KEY.DATA_USE_RESTRICTION,
    width: { max: "1fr", min: "146px" },
  });
```

The `viewBuilder` `V.buildDataUseRestrictions` should probably look something like this:

```
export const buildAggregatedDataUseRestriction = (
  filesResponse: FilesResponse
): React.ComponentProps<typeof C.BasicCell> => {
  return {
    value: processEntityArrayValue(
      filesResponse.projects,
      "dataUseRestriction"
    ),
  };
};
```

Confirm on the `files` tab that you can see in the `Data Use Restriction` column, data such as "NRES" or "Unspecified".

image

File: site-config/hca-dcp/cc-ma-dev/index/projectsEntityConfig.ts [hunk 3]
```diff
@@ -33,8 +34,11 @@ export function getMAEntitiesConfig(
...
   return [...entities].map((entity) => {
-    if (entity.route !== "projects") {
-      return entity;
+    if (entity.route === "projects") {
+      return getMAProjectsEntityConfig(entity);
+    }
+    if (entity.route === "files") {
+      return getMAFilesEntityConfig(entity);
     }
     return getMAProjectsEntityConfig(entity);
```
Comment (frano-m) [inline site-config/hca-dcp/cc-ma-dev/index/projectsEntityConfig.ts]: I believe the default returned value, will always be `projects` related entity config. This means on the `samples` tab the column config is now `projects` related.

I think the default return value on line `43` should be just `entity`.

**Review by frano-m — CHANGES_REQUESTED** [review]
Hi @jpaten would you mind taking a look at each comment and resolving please! Thank you!

File: app/viewModelBuilders/azul/hca-dcp/common/viewModelBuilders.ts (hunk 1 above)
Comment (mrtopsyt) [reply app/viewModelBuilders/azul/hca-dcp/common/viewModelBuilders.ts:750]: Hi @frano-m! I saw that, but the ticket specified that null values should be "" (https://github.com/DataBiosphere/data-browser/issues/4136), so I used `LABEL.EMPTY`. Let me know if you still want me to change that.

File: site-config/hca-dcp/cc-ma-dev/index/projectsEntityConfig.ts (hunk 3 above)
Comment (mrtopsyt) [reply site-config/hca-dcp/cc-ma-dev/index/projectsEntityConfig.ts]: Thanks! Not sure how I missed that one

File: app/viewModelBuilders/azul/hca-dcp/common/viewModelBuilders.ts (hunk 1 above)
Comment (MillenniumFalconMechanic) [reply app/viewModelBuilders/azul/hca-dcp/common/viewModelBuilders.ts:750]: Hi @jpaten, yes, keep as empty string (as per [Slack](https://clevercanary.slack.com/archives/C07V99P98HK/p1732785330654209?thread_ts=1732657487.611409&cid=C07V99P98HK)).

File: app/apis/azul/common/utils.ts
Comment (frano-m) [inline app/apis/azul/common/utils.ts:1]: Let's leave line 144 as is:

`return processNullElements(values, label) ?? [label];`

However, I think the `?? [label]` is probably no longer required?

File: app/viewModelBuilders/azul/hca-dcp/common/viewModelBuilders.ts [hunk 4]
```diff
@@ -188,6 +188,25 @@ export const buildAggregatedDateLastModifiedDate = (
...
 };
 
+/**
+ * Build props for data use restriction BasicCell component from the given files response.
+ * @param filesResponse - Response model return from files API.
+ * @returns model to be used as props for the BasicCell component
+ */
+export const buildAggregatedDataUseRestriction = (
+  filesResponse: FilesResponse
+): React.ComponentProps<typeof C.BasicCell> => {
```
Comment (frano-m) [inline app/viewModelBuilders/azul/hca-dcp/common/viewModelBuilders.ts]: Sorry @jpaten - I might have given you incorrect suggestion for the `buildAggregatedDataUseRestriction`. I wonder if we should make this an `NTagCell` and return all possible values from the `filesResponse`?

If so, update:
- lines 192 and 194 with `NTagCell` in the comments.
- line 198 with `C.NTagCell`
- the return values to be an array of values.

File: site-config/hca-dcp/cc-ma-dev/index/filesEntityConfig.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  const cloneList = { ...cloneEntity.list };
+  // Clone columns.
+  const cloneColumns = [...cloneList.columns];
+  // Add data use restriction column.
+  cloneColumns.splice(2, 0, {
+    ...COLUMN.DATA_USE_RESTRICTION,
+    componentConfig: {
+      component: C.BasicCell,
+      viewBuilder: V.buildAggregatedDataUseRestriction,
+    } as ComponentConfig<typeof C.BasicCell>,
```
Comment (frano-m) [inline site-config/hca-dcp/cc-ma-dev/index/filesEntityConfig.ts]: @jpaten, if we should be displaying all possible `dataUseRestriction` values on the `filesResponse`, let's change the component on lines 27 and 29 to `C.NTagCell`.

**Review by frano-m — CHANGES_REQUESTED** [review]
Sorry @jpaten I may have given you a bit of misdirection - can you confirm with @MillenniumFalconMechanic or @NoopDog as to whether we should show all possible `dataUseRestriction` values on the `filesResponse`? If so, update as per my comments. Thank you again!

File: app/apis/azul/common/utils.ts
Comment (mrtopsyt) [reply app/apis/azul/common/utils.ts:1]: Yeah that confused me - it was like this before, and I don't think it ever was required?

**Comment by mrtopsyt** [comment]
Hi @frano-m! I've updated the Data Use Restriction cell on Files to use an `NTagCell`. I had one remaining question regarding the filters - I noticed they show "Unspecified" when the displayed value is blank, is that acceptable here?

Screenshot 2024-12-02 at 9 37 54 PM

**Comment by mrtopsyt** [comment]
> Hi @frano-m! I've updated the Data Use Restriction cell on Files to use an `NTagCell`. I had one remaining question regarding the filters - I noticed they show "Unspecified" when the displayed value is blank, is that acceptable here?
> 
> Screenshot 2024-12-02 at 9 37 54 PM

Resolved!

**Review by frano-m — APPROVED** [review]
Perfect @jpaten! Thank you!
